File: android/src/main/kotlin/com/example/zendesk_sdk/result/ZendeskResult.kt
package com.example.zendesk_sdk.result

import io.flutter.plugin.common.MethodChannel

/// Typed native result mirroring Dart [ZendeskResult].
sealed class ZendeskResult<out T> {
    data class Success<T>(val value: T) : ZendeskResult<T>()
    data class Failure(val code: String, val message: String?) : ZendeskResult<Nothing>()

    fun complete(result: MethodChannel.Result) {
        when (this) {
            is Success -> result.success(value)
            is Failure -> result.error(code, message, null)
        }
    }
}

File: ios/zendesk_sdk/Sources/zendesk_sdk/Result/ZendeskResult.swift
import Flutter

/// Typed native result mirroring Dart ZendeskResult.
enum ZendeskNativeResult<T> {
    case success(T)
    case failure(code: String, message: String?)

    func complete(_ result: @escaping FlutterResult) {
        switch self {
        case .success(let value):
            result(value)
        case .failure(let code, let message):
            result(FlutterError(code: code, message: message, details: nil))
        }
    }
}
